ecommerce/models.py

```diff
@@ -0,0 +1,274 @@
+    price = models.DecimalField(decimal_places=2, max_digits=20)
+    description = models.TextField()
+
+    class Meta:
+        indexes = [models.Index(fields=["created_on"])]
+
+    def __str__(self):
+        """Description of a ProductVersion"""
+        return f"ProductVersion for {self.description}, ${self.price}"
+
+
+class Basket(TimestampedModel):
+    """
+    Represents a User's basket
+    """
+
+    user = models.OneToOneField(settings.AUTH_USER_MODEL, on_delete=models.PROTECT)
+
+    def __str__(self):
+        """Description of Basket"""
+        return f"Basket for {self.user}"
+
+
+class BasketItem(TimestampedModel):
+    """
+    Represents a product in a user's basket
+    """
+
+    product = models.ForeignKey(Product, on_delete=models.PROTECT)
+    basket = models.ForeignKey(
+        Basket, on_delete=models.PROTECT, related_name="basketitems"
+    )
+    quantity = models.PositiveIntegerField()
+
+    def __str__(self):
+        """Description of BasketItem"""
+        return f"BasketItem for {self.quantity} {self.product}(s)"
+
+
+class Order(TimestampedModel, AuditableModel):
+    """
+    An order containing information for a purchase.
+    """
+
+    FULFILLED = "fulfilled"
+    FAILED = "failed"
+    CREATED = "created"
+    REFUNDED = "refunded"
+
+    STATUSES = [CREATED, FULFILLED, FAILED, REFUNDED]
+    FULFILLED_STATUSES = [FULFILLED]
+
+    purchaser = models.ForeignKey(
+        settings.AUTH_USER_MODEL, on_delete=models.PROTECT, related_name="orders"
+    )
+    status = models.CharField(
+        choices=[(status, status) for status in STATUSES],
+        default=CREATED,
+        max_length=30,
+        db_index=True,
+    )
+
+    def __str__(self):
+        """Description for Order"""
+        return f"Order #{self.id}, status={self.status}"
+
+    @classmethod
+    def get_audit_class(cls):
+        return OrderAudit
+
+    def to_dict(self):
+        """
+        Get a serialized representation of the Order and any attached Basket and Lines
+        """
+        raise NotImplementedError
+
+
+class OrderAudit(AuditModel):
+    """
+    Audit model for Order
+    """
+
+    order = models.ForeignKey(Order, null=True, on_delete=models.PROTECT)
+
+    @classmethod
+    def get_related_field_name(cls):
+        return "order"
+
+
+class Line(TimestampedModel):
+    """
+    A line in an Order. This contains information about a specific item which is purchased.
+    """
+
+    order = models.ForeignKey(Order, on_delete=models.PROTECT, related_name="lines")
+    product_version = models.ForeignKey(ProductVersion, on_delete=models.PROTECT)
+    quantity = models.PositiveIntegerField()
+
+    def __str__(self):
+        """Description for Line"""
+        return f"Line for order #{self.order.id}, {self.quantity} {self.product_version}(s)"
+
+
+class CouponInvoice(TimestampedModel):
+    """
+    Model for a coupon. This stores the discount for the coupon and other information about how it should be redeemed.
```

Comment:
Just in general it would help to have a clearer explanation of the specific purposes of `CouponInvoice`, `CouponInvoiceVersion`, `Coupon`, and `CouponVersion`. Here in particular it seems like the docstring is misleading. This model only has one field (`tag`), so it doesn't seem to stores the discount for the coupon and other information about how it should be redeemed.

Reply:
I'll update the docstring to remove that misleading text. I would like the details to go in the RFC rather than in the docstring. Does that make sense?

Reply:
Anything you can add to these docstrings, however brief, would be helpful. As it stands, I'm not able to draw clear lines between these models and the concepts outlined in the [RFC](https://github.com/mitodl/mitxpro/pull/52).

Reply:
Sure, I'll see what I can add